services/dedup/badger/badger.go

```diff
@@ -171,6 +173,53 @@ func (d *BadgerDB) gcLoop() {
 	}
 }
 
+type Dedup struct {
+	badgerDB *BadgerDB
+	cacheMu  sync.Mutex
+	cache    map[string]int64
+}
+
+func (d *Dedup) Get(kv types.KeyValue) (bool, int64, error) {
+	d.cacheMu.Lock()
+	defer d.cacheMu.Unlock()
+	if previous, found := d.cache[kv.Key]; found {
+		return false, previous, nil
+	}
+	previous, found, err := d.badgerDB.Get(kv.Key)
```

Comment:
@cisse21 I might be missing something but wouldn't holding a lock during DB operations like this affect other workers in the processor pool?
Why not hold the lock just to check the cache, then release it and do a get to allow other workers to use the database?

Reply:
```
	err := d.badgerDB.init()
	if err != nil {
		return false, 0, err
	}
	d.cacheMu.Lock()

	if previous, found := d.cache[kv.Key]; found {
		return false, previous, nil
	}
	d.cacheMu.Unlock()

	previous, found, err := d.badgerDB.Get(kv.Key)
	if err != nil {
		return false, 0, err
	}
	d.cacheMu.Lock()
	if !found {
		d.cache[kv.Key] = kv.Value
	}
	d.cacheMu.Unlock()
	return !found, previous, nil
```

Do you mean something like this?

Reply:
@cisse21 yes although in your example you're not unlocking when found. also, maybe we could check the cache again after the `Get` like:

```go
func (d *Dedup) Get(kv types.KeyValue) (bool, int64, error) {
	err := d.badgerDB.init()
	if err != nil {
		return false, 0, err
	}

	d.cacheMu.Lock()
	previous, found := d.cache[kv.Key]
	d.cacheMu.Unlock()
	if found {
		return false, previous, nil
	}

	previous, found, err = d.badgerDB.Get(kv.Key)
	if err != nil {
		return false, 0, err
	}

	d.cacheMu.Lock()
	defer d.cacheMu.Unlock()

	// check the cache again in case the key was set while we were waiting for the lock
	cachePrev, cacheFound := d.cache[kv.Key]
	if cacheFound { // it's in the cache after all, discarding the value we just read
		return false, cachePrev, nil
	}
	if found { // still not in the cache, but it's in the DB so let's refresh the cache
		d.cache[kv.Key] = kv.Value
	}

	return !found, previous, nil
}
```